gc.c

```diff
@@ -1213,9 +1207,15 @@ rb_data_free(void *objspace, VALUE obj)
     return true;
 }
 
+static void obj_free_object_id(VALUE obj);
+
 void
 rb_gc_obj_free_vm_weak_references(VALUE obj)
 {
+    if (FL_TEST_RAW(obj, FL_SEEN_OBJ_ID)) {
```

Comment:
`FL_SEEN_OBJ_ID` isn't very useful anymore, except for `SHAPE_TOO_COMPLEX`, but we could improve that too.

If we always check the shape flags, we could reclaim that flag.

Also more broadly, `shape_id_t` is `32bit` on 64bit systems, which is way more than we need. If needing to lookup the shape to know if it has an `object_id` (or is frozen) we could actually pack these tags in the most significant bits of the `shape_id`.